relayer/workers/ethrelayer/main.go

```diff
@@ -1,160 +1,147 @@
 	// channel for headers from ethereum (it's a blocking channel so that we
 	// can guarantee that a header is forwarded before we send dependent messages)
 	ethHeaders := make(chan chain.Header)
 
-	ethereumListener, err := NewEthereumListener(ethereumConfig, ethereumConn, ethMessages, ethHeaders, log)
+	listener := NewEthereumListener(
+		w.ethconfig,
+		w.ethconn,
+		ethMessages,
+		ethHeaders,
+		w.log,
+	)
+	writer := NewParachainWriter(
+		w.paraconn,
+		ethMessages,
+		ethHeaders,
+		w.log,
+	)
+
+	finalizedBlockNumber, err := w.queryFinalizedBlockNumber()
 	if err != nil {
-		return nil, err
+		return err
 	}
+	w.log.WithField("blockNumber", finalizedBlockNumber).Debug("Retrieved finalized block number from parachain")
 
-	parachainWriter, err := NewParachainWriter(parachainConn, ethMessages, ethHeaders, log)
+	err = listener.Start(ctx, eg, finalizedBlockNumber+1, uint64(w.ethconfig.DescendantsUntilFinal))
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &Worker{
-		ethereumConfig:   ethereumConfig,
-		ethereumConn:     ethereumConn,
-		parachainConfig:  parachainConfig,
-		parachainConn:    parachainConn,
-		ethereumListener: ethereumListener,
-		parachainWriter:  parachainWriter,
-		log:              log,
-	}, nil
+	err = writer.Start(ctx, eg)
+	if err != nil {
+		return err
+	}
 
+	return nil
 }
 
-func (worker *Worker) Start(ctx context.Context, eg *errgroup.Group) error {
-	worker.log.Info("Starting worker")
-
-	if worker.ethereumListener == nil || worker.parachainWriter == nil {
-		return fmt.Errorf("Sender and/or receiver need to be set before starting chain")
+func (w *Worker) queryFinalizedBlockNumber() (uint64, error) {
+	storageKey, err := types.CreateStorageKey(w.paraconn.Metadata(), "VerifierLightclient", "FinalizedBlock", nil, nil)
+	if err != nil {
+		return 0, err
 	}
 
-	err := worker.ethereumConn.Connect(ctx)
+	var finalizedHeader ethereum.HeaderID
+	_, err = w.paraconn.Api().RPC.State.GetStorageLatest(storageKey, &finalizedHeader)
 	if err != nil {
-		return err
+		return 0, err
 	}
 
-	// Short-lived channels that communicate initialization parameters
-	// between the two chains. The chains close them after startup.
-	ethInit := make(chan chain.Init)
-
-	eg.Go(func() error {
-		ethInitHeaderID := (<-ethInit).(*ethereum.HeaderID)
-		worker.log.WithFields(logrus.Fields{
-			"blockNumber": ethInitHeaderID.Number,
-			"blockHash":   ethInitHeaderID.Hash.Hex(),
-		}).Debug("Received init params for Ethereum from Substrate")
-
-		if worker.ethereumListener != nil {
-			err = worker.ethereumListener.Start(ctx, eg, uint64(ethInitHeaderID.Number),
-				uint64(worker.ethereumConfig.DescendantsUntilFinal))
-			if err != nil {
-				return err
-			}
-		}
-
-		return nil
-	})
+	return uint64(finalizedHeader.Number), nil
+}
 
-	err = worker.parachainConn.Connect(ctx)
+func (w *Worker) connect(ctx context.Context) error {
+	kpForEth, err := secp256k1.NewKeypairFromString(w.ethconfig.PrivateKey)
```

Comment:
dont need this, do we?

Reply:
It's used below. Apologies for the messy diff. I took the opportunity to simplify this worker a lot

Reply:
yeh, i mean it can be removed and then just use nil below, as the eth connection never sends any transactions, only queries state?

Reply:
ah yes. i'll remove